iceoryx_posh/include/iceoryx_posh/experimental/popo/base_publisher.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_EXPERIMENTAL_POSH_POPO_BASE_PUBLISHER_HPP
+#define IOX_EXPERIMENTAL_POSH_POPO_BASE_PUBLISHER_HPP
+
+#include "iceoryx_posh/experimental/popo/sample.hpp"
+#include "iceoryx_posh/internal/popo/ports/publisher_port_user.hpp"
+
+namespace iox {
+namespace popo {
+
+using uid_t = uint64_t;
+
+///
+/// @brief The Publisher class defines the publisher interface used by the Sample class to make it generic.
+/// This allows any publisher specialization to be stored as a reference by the Sample class.
+///
+template<typename T>
+class PublisherInterface
+{
+public:
+    virtual void publish(Sample<T>& sample) noexcept = 0;
+protected:
+    PublisherInterface() = default;
+};
+
+
+///
+/// @brief The BasePublisher class contains the common implementation for the different publisher specializations.
+///
+template<typename T, typename port_t = PublisherPortUser>
+class BasePublisher : public PublisherInterface<T>
+{
+protected:
+
+    BasePublisher(const BasePublisher& other) = delete;
+    BasePublisher& operator=(const BasePublisher&) = delete;
+    BasePublisher(BasePublisher&& rhs) = default;
+    BasePublisher& operator=(BasePublisher&& rhs) = default;
+    ~BasePublisher() = default;
+
+    ///
+    /// @brief uid Get the UID of the publisher.
+    /// @return The publisher's UID.
+    ///
+    uid_t uid() const noexcept;
+
+    ///
+    /// @brief loan Get a sample from loaned shared memory.
+    /// @param size The expected size of the sample.
+    /// @return An instance of the sample that resides in shared memory or an error if unable ot allocate memory to laon.
+    /// @details The loaned sample is automatically released when it goes out of scope.
+    ///
+    cxx::expected<Sample<T>, AllocationError> loan(uint32_t size) noexcept;
```

Comment:
In PR #275 a discussion was going on and the outcome was that all expected/optional methods should be named `try...` to reflect their possible failure.
As we see here that means a lot of changes and `try`s. Is that still what we want or doesn't expected/optional return type already imply that? These return types can't be used directly  anyway and have to be checked/used with `hasValue()`/`getValue()`.

Reply:
I'm not a fan of prefixing all the methods with `try`.

The `tryX` naming strategy is reminiscent of C programming, where functions may return null and must be explicitly checked before use otherwise they might cause runtime errors. These kind of errors are not picked up in compilation so in that scenario it makes sense.
As mentioned, the `cxx::expected` and `cxx::optional` types are not directly usable so the same issue does does not exist here - doing so will result in a compilation error.

Reply:
IMHO we should be consistent throughout the code base. So seems like adding `tryX` would be a good way forward ;-)

Reply:
I would side with @ithier on this issue. In my opinion the `try` prefix makes sense for methods which are returning an optional. In my mind it is an expression which states: I would like to have a result but if there is none it is not an error or failure. Therefore I'll try to get it.

This method does return an `cxx::expected`. When the return value does not contain a `Sample<T>` there is definitely something wrong which has to be handled immediately! It is not alright to `try` loan a message and say: mmmhhh there is none maybe later. This is alright with `optional` but when an `expected` does not contain a value something is smelling.